[processor/tailsampling] Simplify tsp locking

csmarchbanks opened the pull request:
<!--Ex. Fixing a bug - Describe the bug and how this fixes the issue.
Ex. Adding a feature - Explain what this achieves.-->
#### Description
This PR changes the tailsampling processor to do all work (append traces to a batch, decide on which traces from a batch are sampled, forward to the next processor) in a single goroutine. At Grafana Labs we have found that there is significant contention between the various mutexes such that little performance is gained by allowing traces to be updated while a policy evaluation is occurring. In addition, the locking adds a lot of complexity which makes future improvements more challenging, and introduces subtle concurrency issues such as seen in the tracking issue linked below.

We (mostly @Logiraptor) also experimented with using multiple workers, each with their own goroutine to process traces but we found little improvement as yet again various contention issues showed up. See https://github.com/Logiraptor/opentelemetry-collector-contrib/pull/1 (or the first commit in this PR) for more details.

If additional performance is necessary for slow decision tick latencies I would suggest adding parallelism just to the call to `makeDecision`, which would allow additional CPUs to be used without adding contention on `idToTrace` or other locations.

<!-- Issue number (e.g. #1234) or full URL to issue, if applicable. -->
#### Link to tracking issue
Fixes #41656 
Working towards https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/43876

<!--Describe what testing was performed and which tests were added.-->
#### Testing
Tests were updated to make sure the tick behavior remains correct and consistent. In addition, the syncBatcher needed to be updated to properly implement `Stop`.

Logiraptor committed:
[tailsamplingprocessor] Parallel decision ticker

This PR builds on #43201 (pending) to parallelize the decision ticker.
It also fixes the open issue in #41656 by moving all state management to
a single goroutine (per shard).

In practice at Grafana Labs, we've seen that the single threaded nature
of the decision ticker is the primary bottleneck when running the tail
sampling processor at high scales. For example, we have one large
cluster performing around 12M policy evaluations per second while
maintaining p99 tick latency of around 500ms. In order to reach this
scale, we need to run ~200 replicas just to get enough parallel threads
of execution. By adding support for parallel decision ticks, we hope to
run fewer, larger replicas which should reduce the fragmentation
caused by trace id load balancing.

Another benefit here is to reduce contention on the idToTrace map and generally
simplify reasoning about concurrency by giving each trace a clear
"owner" goroutine. Previously we'd try to do some work inside the calling
goroutine of ConsumeTraces. Instead, new work is passed via channels to
its dedicated shard. By default we run one shard per available core.
Shared state is kept to the minimum required to implement TSP semantics:

1. The trace limiter is shared, which allows the overall memory
   consumption to be reliably capped even in case of uneven shard
   utilization.
2. Both decision caches are shared, which again allows overall memory
   consumption to be capped even with uneven shard utilization.

In practice, shard utilization should be very even, but it's still nice
to have hard guarantees on the total traces in memory, and this allows
us to avoid changing any tests in this PR, which greatly increases my
confidence in the code despite the large change. Ideally the change to
parallel execution is invisible to end users (aside from the much higher throughput).

Sharding makes use of the maphash package with a unique `Seed`, which
should provide even shard distribution even if traceIDs have previously
been sampled downstream. This is better than e.g. a simple modulo on
traceID, since traceIDs are not necessarily uniformly distribute after
head sampling.

csmarchbanks committed:
Remove workers

In testing parallelizing across workers did not show much improvement,
so simplify to run everything in only one go routine.

csmarchbanks committed:
Remove unnecessary lock on trace structure

Now that all evaluations will be done in the same goroutine as receiving
data it is no longer necessary to lock each trace during evaluation as
the trace will not be modified at the same time.

csmarchbanks committed:
Run make gotidy

csmarchbanks committed:
Send full batches of traces to the work channel

This will reduce the amount of small channel sends that we do and makes
sure that we process an entire batch of new traces before doing a policy
tick for a stored batch.

csmarchbanks committed:
Do not discard traces during shutdown

This avoids dropping data during shutdown of the tailsamplingprocessor.
The old behavior may be desired for some users so a flag is left in that
allows dropping the data still, such as if the data will be replayed.

csmarchbanks committed:
Update generated code

csmarchbanks committed:
Add changelog entry

Logiraptor commented on processor/tailsamplingprocessor/processor_test.go:
```diff
@@ -66,49 +65,55 @@ func (t *TestPolicyEvaluator) Evaluate(ctx context.Context, traceID pcommon.Trac
 type testTSPController struct {
-	// tickBarrier is a barrier to block ticks until the test is ready
-	tickBarrier chan struct{}
+	tickChans []chan chan struct{}
```
I think this can just be a `chan chan struct` now. The slice was for having multiple workers

Logiraptor reviewed (commented):
Looks great! Just some minor points of cleanup

csmarchbanks committed:
Remove now unused tracelimiter package

csmarchbanks committed:
Remove slice of tickChans in tests

csmarchbanks committed:
Refactor SpanCount to be an int64 instead of atomic

csmarchbanks commented:
Thanks! Updated and expanded the changelog to mention that `SpanCount` changed as well.

csmarchbanks committed:
Merge branch 'main' into simplify-tsp-locking

csmarchbanks committed:
Update comment that no longer has a lock to reference

atoulme commented:
any performance benchmark we can use to track the impact of this work?

atoulme commented on processor/tailsamplingprocessor/processor_benchmarks_test.go:
can you please do this in a separate PR and open an issue to fix it?

csmarchbanks committed:
Add deleted benchmark back

csmarchbanks committed:
Avoid potential int overflow

csmarchbanks committed:
Revert unnecessary test change

csmarchbanks commented on processor/tailsamplingprocessor/config.go:
```diff
@@ -291,4 +291,10 @@ type Config struct {
 	SampleOnFirstMatch bool `mapstructure:"sample_on_first_match"`
+	// WorkChanCapacity sets the capacity of the work channel. This is the number of batches of traces to store in the
+	// channel, not individual traces.
+	WorkChanCapacity int `mapstructure:"work_chan_capacity"`
```
I'll definitely add DropPendingTracesOnShutdown to the README.

Thinking about it some more I wonder if we should just have a reasonable default to help with throughput and no config unless someone reports they need it. What do you think?

csmarchbanks committed:
Add benchmark for processor throughput

csmarchbanks committed:
Group traces before sending to channel

This improves throughput by doing the work that does not depend on a
lock before moving to the single threaded part of the processor.

csmarchbanks committed:
Use more realistic policies for benchmark

csmarchbanks committed:
Add throughput in bytes to benchmark

csmarchbanks committed:
Fix lint errors

csmarchbanks commented:
In addition, as I mentioned in the description, we could parallelize the decision part of the code if desired. A very rough implementation of that reports 1.3ms/op or 215MB/s, so faster than the implementation from main.

csmarchbanks committed:
Remove configurable work channel size

csmarchbanks committed:
Add new config option to readme

csmarchbanks committed:
Move new benchmark to benchmark file

csmarchbanks committed:
Move copying trace data to ConsumeTraces

Moving the code that copies all the data from inside of the TSP loop to
before ingest greatly improves throughput by allowing it to be
parallelized before the single threaded inner loop. The cost of copying
is paid in almost all scenarios so this should be an overall
improvement. The only scenario where more work is being done is when
there was a cached decision to not sample a trace.

csmarchbanks committed:
Merge branch 'main' into simplify-tsp-locking

csmarchbanks committed:
Merge remote-tracking branch 'origin/main' into simplify-tsp-locking

csmarchbanks committed:
Merge remote-tracking branch 'origin/main' into simplify-tsp-locking

csmarchbanks committed:
Merge remote-tracking branch 'origin' into simplify-tsp-locking

csmarchbanks committed:
Merge branch 'main' into simplify-tsp-locking

atoulme commented on processor/tailsamplingprocessor/internal/sampling/span_count_sampler.go:
```diff
@@ -34,11 +34,11 @@ func NewSpanCount(settings component.TelemetrySettings, minSpans, maxSpans int32
 	c.logger.Debug("Evaluating spans counts in filter")
 
-	spanCount := int(traceData.SpanCount.Load())
+	spanCount := int32(traceData.SpanCount)
```
This can still overflow. Using int32 can truncate the number. Here is an example: https://go.dev/play/p/qL-jkVoRjGq

csmarchbanks replied:
I noticed a couple places in processor.go so fixed those as well.

csmarchbanks committed:
Avoid a couple other int64 -> int conversions

csmarchbanks committed:
Merge branch 'main' into simplify-tsp-locking
